src/webgpu/api/validation/layout_shader_compat.spec.ts

```diff
@@ -1,14 +1,318 @@
+    switch (bindableResource) {
+      case 'compareSamp':
+        return 'var tmp : sampler_comparison';
+      case 'filtSamp':
+      case 'nonFiltSamp':
+        return 'var tmp : sampler';
+      case 'sampledTex':
+        return 'var tmp : texture_2d<f32>';
+      case 'sampledTexMS':
+        return 'var tmp : texture_multisampled_2d<f32>';
+      case 'storageBuf':
+        return 'var<storage, read_write> tmp : vec4u';
+      case 'readonlyStorageBuf':
+        return 'var<storage, read> tmp : vec4u';
+      case 'uniformBuf':
+        return 'var<uniform> tmp : vec4u;';
+      case 'readonlyStorageTex':
+        return 'var tmp : texture_storage_2d<r32float, read>';
+      case 'writeonlyStorageTex':
+        return 'var tmp : texture_storage_2d<r32float, write>';
+      case 'readwriteStorageTex':
+        return 'var tmp : texture_storage_2d<r32float, read_write>';
+    }
+  }
+}
+
+export const g = makeTestGroup(F);
+
+g.test('pipeline_layout_shader_exact_match')
+  .desc(
+    `
+  Test that the binding type in the pipeline layout must match the related declaration in shader.
+  Note that read-write storage textures in the pipeline layout can match write-only storage textures
+  in the shader.
+  `
+  )
+  .params(u =>
+    u
+      .combine('bindingInPipelineLayout', kBindableResources)
+      .combine('bindingInShader', kBindableResources)
+      .beginSubcases()
+      .combine('pipelineLayoutVisibility', kShaderStageCombinations)
+      .combine('shaderStageWithBinding', kShaderStages)
+      .combine('isBindingStaticallyUsed', [true, false] as const)
+      .unless(
+        p =>
+          // We don't test using non-filtering sampler in shader because it has the same declaration
+          // as filtering sampler.
+          p.bindingInShader === 'nonFiltSamp' ||
+          ((p.bindingInPipelineLayout === 'writeonlyStorageTex' ||
+            p.bindingInPipelineLayout === 'readwriteStorageTex' ||
+            p.bindingInPipelineLayout === 'storageBuf') &&
+            (p.pipelineLayoutVisibility & GPUConst.ShaderStage.VERTEX) > 0) ||
+          ((p.bindingInShader === 'writeonlyStorageTex' ||
+            p.bindingInShader === 'readwriteStorageTex' ||
+            p.bindingInPipelineLayout === 'storageBuf') &&
+            p.shaderStageWithBinding === GPUConst.ShaderStage.VERTEX)
+      )
+  )
+  .fn(t => {
+    const {
+      bindingInPipelineLayout,
+      bindingInShader,
+      pipelineLayoutVisibility,
+      shaderStageWithBinding,
+      isBindingStaticallyUsed,
+    } = t.params;
+
+    const layout = t.createPipelineLayout(bindingInPipelineLayout, pipelineLayoutVisibility);
+    const bindResourceDeclaration = `@group(0) @binding(0) ${t.GetBindableResourceShaderDeclaration(
+      bindingInShader
+    )}`;
+    const staticallyUseBinding = isBindingStaticallyUsed ? '_ = tmp; ' : '';
+    const isAsync = false;
+    const success =
+      !isBindingStaticallyUsed ||
```

Comment:
nit: Could we break down this massive expression in subexpressions with names that explain what's going on? Or something like:

```
var success = true;

if (isBindingStaticallyUsed) {
  success = bindingInPipelineLayout === bindingInShader;

  // Filtering and non-filtering both have the same shader declaration.
  success |= bindingInPipelineLayout === 'nonFiltSamp' && bindingInShader === 'filtSamp';

  // Promoting storage textures that are read-write in the layout can be readonly in the shader.
 success |= bindingInPipelineLayout === 'readwriteStorageTex' && bindingInShader === 'writeonlyStorageTex';
 
  // The shader using the resource must be included in the visibility in the layout.
  success &= pipelineLayoutVisibility & shaderStageWithBinding;
}
```